src/1-theory/10-async.adoc

```diff
@@ -0,0 +1,407 @@
+seconds. You realize that the payment gateway is too slow to use synchronous
+requests to talk to them, so lets shift that to a worker.
+
+.Letting the API do a bit less work up front, shifting the waiting work to a job queue, and letting the GUI short poll for updates.
+[plantuml]
+----
+@startuml
+
+title Accepting Payments: Short Polling and Job Queues
+
+boundary "Web GUI" as GUI
+control "API"
+control Worker
+boundary "Payment Gateway" as PG
+
+GUI -> API: POST /payments
+API -> Worker: Queue Job
+GUI <-- API: HTTP 201 Created (with a URL to poll)
+GUI -> GUI: Sets a timer to start calling the API every X seconds
+
+loop
+    alt payment gateway not ready
+        GUI -> API: GET /payments/abc
+        GUI <-- API: HTTP 200 OK { status: pending }
+    else payment gateway has approved the payment
+        Worker -> PG: POST /submitPaymentAttempt
+        PG -> PG: May take 30-60s to confirm
+        Worker <-- PG: HTTP 200 OK
+        Worker -> API: Update payment status
+        GUI -> API: GET /payments/abc
+        GUI <-- API: HTTP 200 OK { status: approved }
+    end
+end
+
+@enduml
+----
+
+Alright, better. Our client is a bit chatty and is generating a lot of overheard, but hey they are getting relatively real time updates. Maybe this is good enough, but maybe we want to improve this even more with a WebSocket.
+
+.Removing the short polling from the client using a WebSocket.
+[plantuml]
+----
+@startuml
+
+title Accepting Payments: WebSocket and Job Queues
+
+boundary "Web GUI" as GUI
+control "API"
+control Worker
+boundary "Payment Gateway" as PG
+
+GUI -> API: POST /payments
+API -> Worker: Queue Job
+GUI <-- API: HTTP 201 Created (with a wss:// URL)
+GUI -> API: Establish wss:// connection
+Worker -> PG: POST /submitPaymentAttempt
+PG -> PG: May take 30-60s to confirm
+Worker <-- PG: HTTP 200 OK
+Worker -> API: Update payment status
+GUI <-- Worker: event triggered: payment being attempted now
+GUI <-- Worker: event triggered: payment success
+
+@enduml
+----
+
+Now the client is getting real time updates, and not just "is it done or not" but actual progress on various things happening. Also the background worker is able to talk directly to the client, instead of just updating the state in the API and hoping the client checks for a change.
+
+Again this might be just fine, but the last remaining bottleneck is the interaction with this slow third party API. Even though the slowness is in the background worker, maybe the worker is backing up a bunch, due to high demand. During Black Friday this might mean your workers are so backed up they're taking 20 minutes to get to updating, because most of them are spending all the CPU cycles just waiting.
+
+Maybe the company providing this payment gateway realized that 30-60 sync requests are daft, and implemented Web Hooks on their end. Let's update our approach to support that.
+
+.Removing the short polling from the client using a WebSocket.
```

Comment:
Title leftover from the previous example?

Reply:
The title before it mentioned short polling, i think this is consistent.